src/clipboardobserver.js

```diff
@@ -30,28 +30,72 @@ export default class ClipboardObserver extends DomEventObserver {
 		function handleInput( evt, data ) {
 			data.preventDefault();
 
-			doc.fire( 'clipboardInput', { dataTransfer: data.dataTransfer } );
+			const targetRanges = data.dropRange ? [ data.dropRange ] : Array.from( doc.selection.getRanges() );
+
+			doc.fire( 'clipboardInput', {
+				dataTransfer: data.dataTransfer,
+				targetRanges
```

Comment:
I looked into the code and I still think that `clipboardInput` is quite useful. It's where all the handling should be. Even the image upload feature uses this event. If not for it, you'd need to add 2 listeners. The same in the clipboard feature itself – it'd need two listeners.

The only thing is that we want to select the whole content after drop happened. But this is far down inside `inputTransformation` which was meant to be that "single pipe" anyway, so getting rid of `clipboardInput` won't help with that at all.

One thing which I don't like here is that `inputTransformation` does content insertion which is wrong. So I think that this requires a bigger refactoring anyway so it's not a part of this ticket anyway.